### PR TITLE
driver: crypto: hisilicon: add ECC gen_keypair and ECDH

### PR DESCRIPTION
**Pull request by yuzexiyzx**
add ECC gen_keypair and shared secret

### CODE REVIEW

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 1]
```diff
@@ -0,0 +1,997 @@
...
+
+	if (curve->id == TEE_ECC_CURVE_SM2)
+		msg->sm2_sp = true;
+
+	msg->curve_bytes = BITS_TO_BYTES(curve->key_bits);
+	msg->key_bytes = hpre_ecc_get_hw_kbytes(curve->key_bits);
+	if (!msg->key_bytes)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	ret = hpre_ecc_dh_params_alloc(msg);
```
Comment (anilreddy3) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:805]: Type mismatch here, hpre_ecc_dh_params_alloc returns of type enum hisi_drv_status which is assigned to type TEE_Result.
and i see enum hisi_drv_status declaration is missing in this patch.

Comment (yuzexiyzx) [reply core/drivers/crypto/hisilicon/hpre_ecc.c:806]: solved. enum hisi_drv_status declaration was merged in patch about qm.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 2]
```diff
@@ -0,0 +1,998 @@
...
+	0x02, 0xdf, 0x32, 0xe5, 0x21, 0x39, 0xf0, 0xa0
+};
+
+static uint8_t g_sm2_n[] = {
+	0xff, 0xff, 0xff, 0xfe, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+	0xff, 0xff, 0xff, 0xff, 0x72, 0x03, 0xdf, 0x6b, 0x21, 0xc6, 0x05, 0x2b,
+	0x53, 0xbb, 0xf4, 0x09, 0x39, 0xd5, 0x41, 0x23
+};
+
+static struct hpre_ecc_curve g_curve_list[] = {
```
[the remark this answers is not in the document]
Comment (yuzexiyzx) [reply core/drivers/crypto/hisilicon/hpre_ecc.c]: image
It seems that this could not be const.

**Comment by yuzexiyzx** [comment]
@etienne-lms All comments have been solved.

**Comment by yuzexiyzx** [comment]
@etienne-lms All comments have been solved.

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 3]
```diff
@@ -0,0 +1,995 @@
...
+		return HISI_QM_DRVCRYPT_EINVAL;
+
+	memcpy(b, curve->b, msg->curve_bytes);
+
+	ecc_dh->x_bytes = msg->curve_bytes;
+	ecc_dh->y_bytes = msg->curve_bytes;
+	if (!pubkey) { /* gen key pair */
+		memcpy(x, curve->x, ecc_dh->x_bytes);
+		memcpy(y, curve->y, ecc_dh->y_bytes);
+	} else { /* do shared secret */
```
Comment (etienne-lms) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:742]: prefer:
```c
	} else {
		/* Do shared secret */

File: core/drivers/crypto/hisilicon/hpre_ecc.c [hunk 4]
```diff
@@ -0,0 +1,995 @@
...
+		ecc_dh->y_bytes = crypto_bignum_num_bytes(pubkey->y);
+	}
+
+	ret = hpre_ecc_dh_transfer_key(msg);
+	if (ret)
+		return ret;
+
+	ret = hpre_ecc_dh_transfer_in(msg);
+
+	return ret;
```
Comment (etienne-lms) [inline core/drivers/crypto/hisilicon/hpre_ecc.c:755]: nitpicking: `return hpre_ecc_dh_transfer_in(msg);`

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@foss.st.com>` with comments addressed.

**Comment by yuzexiyzx** [comment]
@jforissier All comments have been solved.

**Pull request merged**